designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/ProcessActionRepository.scala

```diff
@@ -26,46 +29,105 @@ trait ProcessActionRepository[F[_]] {
-    for {
+  def addInstantAction(processId: ProcessId, processVersion: VersionId, actionType: ProcessActionType, comment: Option[Comment], buildInfoProcessingType: Option[ProcessingType])
+                      (implicit user: LoggedUser): F[ProcessActionEntityData] = {
+    val now = Instant.now()
+    run(for {
       commentId <- newCommentAction(processId, processVersion, comment)
-      processActionData = ProcessActionEntityData(
-        processId = processId,
-        processVersionId = processVersion,
-        user = user.username,
-        performedAt = Timestamp.from(Instant.now()),
-        action = action,
-        commentId = commentId,
-        buildInfo = buildInfo
-      )
-      _ <- processActionsTable += processActionData
-    } yield processActionData
+      result <- insertAction(None, processId, Some(processVersion), actionType, ProcessActionState.Finished, now, Some(now), commentId, buildInfoProcessingType)
+    } yield result)
+  }
+
+  private def insertAction(actionIdOpt: Option[ProcessActionId], processId: ProcessId, processVersion: Option[VersionId], actionType: ProcessActionType, state: ProcessActionState,
+                           createdAt: Instant, performedAt: Option[Instant], commentId: Option[Long], buildInfoProcessingType: Option[ProcessingType])
+                          (implicit user: LoggedUser): DB[ProcessActionEntityData] = {
+    val actionId = actionIdOpt.getOrElse(ProcessActionId(UUID.randomUUID()))
+    val buildInfoJsonOpt = buildInfoProcessingType.flatMap(buildInfos.forType).map(BuildInfo.writeAsJson)
+    val processActionData = ProcessActionEntityData(
+      id = actionId,
+      processId = processId,
+      processVersionId = processVersion,
+      user = user.username,
+      createdAt = Timestamp.from(createdAt),
+      performedAt = performedAt.map(Timestamp.from),
+      action = actionType,
+      state = state,
+      commentId = commentId,
+      buildInfo = buildInfoJsonOpt)
+    (processActionsTable += processActionData).map { insertCount =>
+      if (insertCount != 1)
+        throw new IllegalArgumentException(s"Action with id: $actionId can't be inserted")
+      processActionData
+    }
+  }
+
+  private def updateAction(actionId: ProcessActionId, processId: ProcessId, processVersion: Option[VersionId], state: ProcessActionState, performedAt: Option[Instant], commentId: Option[Long])
+                          (implicit user: LoggedUser): DB[Boolean] = {
+    for {
+      updateCount <- processActionsTable.filter(_.id === actionId)
+        .map(a => (a.performedAt, a.state, a.commentId))
+        .update((performedAt.map(Timestamp.from), state, commentId))
+    } yield updateCount == 1
+  }
+
+  // we use "select for update where false" query syntax to lock the table - it is useful if you plan to insert something in a critical section
+  def lockActionsTable: F[Unit] = {
+    run(processActionsTable.filter(_ => false).forUpdate.result.map(_ => ()))
+  }
```

Comment:
but we lock the whole table?

If adding actions is always preceded whit selecting `process` from DB, maybe we could do select for update on process table and add actions in the same transaction, so only one record is locked?

Reply:
Select for update on process table won't work, because we don't update process table, but only insert action.
There is no other way than lock the whole table if you want to do pessimistic locking with insert (not update).
This transaction is short so IMO this lock won't hurt us. If we would like to avoid this locking, we should replace this solution with optimistic locking, but it would complicate the solution. IMO it is not worth it. WDYT?

Reply:
right, but it can be hacked if process table would be also updated in the same trasaction as adding an action. But lets leave it as it is now, as it no big cost, as you said :)